gusto/model.py

```diff
@@ -0,0 +1,228 @@
+from gusto.linear_solvers import ShallowWaterSolver, CompressibleSolver
+from gusto.state import State
+from gusto.transport_equation import VectorInvariant, AdvectionEquation, SUPGAdvection
+
+
+class Model(object, metaclass=ABCMeta):
+
+    eqn_opts = defaultdict(dict)
+
+    def __init__(self,
+                 mesh,
+                 timestepping=None,
+                 output=None,
+                 parameters=None,
+                 diagnostics=None,
+                 diagnostic_fields=None):
+
+        # store timestepping class
+        self.timestepping = timestepping
+
+        # update physical parameters
+        if parameters is not None:
+            for k, v in parameters.items():
+                self.parameters.__setattr__(k, v)
+
+        # create state
+        self.state = State(mesh=mesh,
+                           vertical_degree=self.vertical_degree,
+                           horizontal_degree=self.horizontal_degree,
+                           family=self.family,
+                           fieldlist=self.fieldlist,
+                           output=output,
+                           diagnostics=diagnostics,
+                           diagnostic_fields=diagnostic_fields)
+
+        # figure out if we're on a sphere
+        try:
+            self.on_sphere = (mesh._base_mesh.geometric_dimension() == 3 and mesh._base_mesh.topological_dimension() == 2)
+        except AttributeError:
+            self.on_sphere = (mesh.geometric_dimension() == 3 and mesh.topological_dimension() == 2)
+
+        #  build the vertical normal and define perp for 2d geometries
+        dim = mesh.topological_dimension()
+        if self.on_sphere:
+            x = SpatialCoordinate(mesh)
+            R = sqrt(inner(x, x))
+            self.state.k = interpolate(x/R, mesh.coordinates.function_space())
+            if dim == 2:
+                outward_normals = CellNormal(mesh)
+                self.state.perp = lambda u: cross(outward_normals, u)
+        else:
+            kvec = [0.0]*dim
+            kvec[dim-1] = 1.0
+            self.state.k = Constant(kvec)
+            if dim == 2:
+                self.state.perp = lambda u: as_vector([-u[1], u[0]])
+
+    @abstractproperty
+    def family(self):
+        pass
+
+    @abstractproperty
+    def vertical_degree(self):
+        pass
+
+    @abstractproperty
+    def horizontal_degree(self):
+        pass
+
+    @abstractproperty
+    def fieldlist(self):
+        pass
+
+    @abstractproperty
+    def parameters(self):
+        pass
+
+    @abstractproperty
+    def linear_solver(self):
+        pass
+
+    @abstractproperty
+    def forcing(self):
+        pass
+
+    @abstractproperty
+    def default_field_equations(self):
+        pass
+
+    @abstractproperty
+    def default_advection_schemes(self):
+        pass
+
+    def setup(self, linear_solver=None, forcing=None, field_equations=None,
+              advection_schemes=None):
```

Comment:
> I don't see how the firedrake.utils.cached_property helps so maybe that's the bit of the puzzle that will help me understand what you mean?

This is just a way of "delaying" evaluation of properties (create on first use, rather than create in `__init__`).

> If I understand correctly you are suggesting I pass the equations, advection schemes, linear solver etc etc into `Model.__init__` and this can be done via some functions like this:

> `ShallowWaterModel = Model(stuff)`

> where stuff is a bunch of default arguments for the field list, linear_solver, forcing etc (i.e. replacing the class properties I currently have).

OK, I see.  The problem is that if I want to poke some part of the initialisation (say linear_solver) then I need a lot of stuff already set up (from the Model).

But actually, I sort of only need `state` to do this setup phase.  So would it make sense to split into two bits:

```python
def ShallowWaterState(mesh, ...):
    return State(...)

def CompressibleEulerState(mesh, ...):
    return State(...)

def ShallowWaterModel(state, ...):
    return Model(state, ...)

state = ShallowWaterState(mesh)

advected_fields = ...
model = ShallowWaterModel(state, advected_fields=advected_fields)

...
```
?